symsrv-fetch.py

```diff
@@ -24,345 +24,468 @@
+                )
+
+        return await asyncio.gather(*tasks)
+
+
+def collect(modules):
+    res = asyncio.run(collect_helper(modules))
+    to_dump = []
+    stats = {"no_pdb": 0, "is_there": 0}
+    for filename, debug_id, code_file, code_id, has_pdb, has_code, is_there in res:
+        if not has_pdb:
+            if is_there:
+                stats["is_there"] += 1
+            else:
+                stats["no_pdb"] += 1
+                log.info(f"No pdb for {filename}/{debug_id}")
+            continue
 
-    if not file_index:
         log.info(
-            'No symbols downloaded: %d considered, '
-            '%d already present, %d in blacklist, %d skipped, %d not found, '
-            '%d not processed ' %
-            (total,
-             existing_count,
-             blacklist_count,
-             skiplist_count,
-             not_found_count,
-             not_processed_count))
-        write_skiplist(skiplist)
-        sys.exit(0)
+            f"To dump: {filename}/{debug_id}, {code_file}/{code_id} and has_code = {has_code}"
+        )
+        to_dump.append((filename, debug_id, code_file, code_id, has_code))
+
+    log.info(f"Collected {len(to_dump)} files to dump")
+
+    return to_dump, stats
+
+
+async def fetch_and_write(output, client, filename, file_id):
+    path = os.path.join(filename, file_id, filename)
+    data = await fetch_file(client, MICROSOFT_SYMBOL_SERVER, path)
+
+    if not data:
+        return False
 
-    # Write an index file
-    with zipfile.ZipFile(args.zip, 'w', zipfile.ZIP_DEFLATED) as z:
+    output_dir = os.path.join(output, filename, file_id)
+    if not os.path.exists(output_dir):
+        os.makedirs(output_dir)
+
+    output_path = os.path.join(output_dir, filename)
+    async with AIOFile(output_path, "wb") as Out:
+        await Out.write(data)
+
+    return True
+
+
+async def fetch_all_helper(output, modules):
+    loop = asyncio.get_event_loop()
+    tasks = []
+    connector = TCPConnector(limit=100, limit_per_host=20)
+    async with ClientSession(
+        loop=loop, timeout=ClientTimeout(total=TIMEOUT), connector=connector
+    ) as client:
+        for filename, debug_id, code_file, code_id, has_code in modules:
+            tasks.append(fetch_and_write(output, client, filename, debug_id))
+            if has_code:
+                tasks.append(fetch_and_write(output, client, code_file, code_id))
+
+        return await asyncio.gather(*tasks)
+
+
+def fetch_all(output, modules):
+    res = asyncio.run(fetch_all_helper(output, modules))
+    res = iter(res)
+    fetched_modules = []
+    for filename, debug_id, code_file, code_id, has_code in modules:
+        fetched_pdb = next(res)
+        if has_code:
+            has_code = next(res)
+        if fetched_pdb:
+            fetched_modules.append((filename, debug_id, code_file, code_id, has_code))
+    return fetched_modules
+
+
+def get_base_data(url):
```

Comment:
You can avoid the helper in here too. It might be worth just making main async and move the asyncio.run outside all so you don't have to repeat it multiple times.